bugwarrior/config/schema.py

```diff
@@ -16,133 +28,149 @@
+            {"scheme": scheme},
+        )
 
-        port = parts['port']
-        if port is not None and int(port) > 65_535:
-            raise pydantic.v1.errors.UrlPortError()
+    # Add a dummy scheme to leverage AnyUrl's validation (host, port, etc.)
+    # then validate as a proper URL
+    url_with_scheme = f"https://{value}"
 
-        user = parts['user']
-        if cls.user_required and user is None:
-            raise pydantic.v1.errors.UrlUserInfoError()
+    # NOTE: is this needed ?
+    try:
+        AnyUrl(url_with_scheme)
+    except Exception as exc:
+        raise PydanticCustomError(
+            "url_parsing", "Invalid URL format: {error}", {"error": str(exc)}
+        ) from exc
 
-        return parts
+    # NOTE: I did not migrate the user_required part, because it does not appear anywhere in the code
+    return value.rstrip("/")
 
 
-# Pydantic complicates the use of sets or lists as default values.
-class ConfigList(frozenset):
-    @classmethod
-    def __get_validators__(cls):
-        yield cls.validate
+NoSchemeUrl = Annotated[str, BeforeValidator(validate_no_scheme_url)]
 
-    @classmethod
-    def validate(cls, value):
-        """Cast ini string to a list of strings"""
+
+def parse_config_list(value: str | list[str]) -> list[str]:
+    """Cast ini string to a list of strings."""
+    if isinstance(value, str):
+        return [
+            item.strip()
+            for item in re.split(r",(?![^{]*})", value.strip())
+            if item != ""
+        ]
+    return list(value)
+
+
+ConfigList = Annotated[list[str], BeforeValidator(parse_config_list)]
+
+
+def _expand_env_vars(value: str | Path) -> str:
+    """Expand $VAR and ${VAR} style environment variables."""
+    return Template(str(value)).safe_substitute(os.environ)
+
+
+ExpandedPath = Annotated[
+    Path,
+    BeforeValidator(_expand_env_vars),
+    AfterValidator(lambda path: path.expanduser()),
+]
+
+
+def _relativize_path(path: Path) -> Path:
+    """Convert path to relative path from cwd, or return absolute path if not possible."""
+    # NOTE: Path.relative_to() raises ValueError if path is not a subpath, unlike
+    # os.path.relpath() which can always compute a relative path using "../".
+    try:
+        return path.relative_to(Path.cwd())
+    except ValueError:
         try:
-            return [
-                item.strip()
-                for item in re.split(",(?![^{]*})", value.strip())
-                if item != ''
-            ]
-        except AttributeError:  # not a string, presumably an iterable
-            return value
+            return path.resolve().relative_to(Path.cwd().resolve())
+        except ValueError:
+            return path.resolve()
 
 
-# HACK https://stackoverflow.com/a/34116756
-class ExpandedPath(type(pathlib.Path())):  # type: ignore
-    @classmethod
-    def __get_validators__(cls):
-        yield cls.validate
+LoggingPath = Annotated[ExpandedPath, AfterValidator(_relativize_path)]
```

Comment:
Do you understand why we're making the path relative here? This seems like a lot of code to maintain a behavior which may not have made any sense in the first place.

Reply:
Removed